descwl_shear_sims/sim.py

```diff
@@ -390,6 +360,163 @@ def make_exp(
     return exp
 
 
+def _draw_objects(
+    image, objlist, shifts, psf, draw_method,
+    coadd_bbox_cen_gs_skypos,
+    rng,
+    shear=None,
+):
+
+    wcs = image.wcs
+    kw = {}
+    if draw_method == 'phot':
+        kw['maxN'] = 1_000_000
+        kw['rng'] = galsim.BaseDeviate(seed=rng.randint(low=0, high=2**30))
+
+    for obj, shift in zip(objlist, shifts):
+
+        if shear is not None:
+            obj = obj.shear(shear)
+            shift = shift.shear(shear)
+
+        # Deproject from u,v onto sphere.  Then use wcs to get to image pos.
+        world_pos = coadd_bbox_cen_gs_skypos.deproject(
+            shift.x * galsim.arcsec,
+            shift.y * galsim.arcsec,
+        )
+
+        image_pos = wcs.toImage(world_pos)
+        local_wcs = wcs.local(image_pos=image_pos)
+
+        convolved_object = get_convolved_object(obj, psf, image_pos)
+
+        stamp = convolved_object.drawImage(
+            center=image_pos, wcs=local_wcs, method=draw_method, **kw
+        )
+
+        b = stamp.bounds & image.bounds
+        if b.isDefined():
+            image[b] += stamp[b]
+
+
+def _draw_bright_objects(
+    image,
+    noise,
+    origin,
+    bmask,
+    band,
+    objlist,
+    shifts,
+    mags,
+    psf,
+    coadd_bbox_cen_gs_skypos,
+    mask_threshold,
+    rng,
+    star_bleeds,
+):
+    # extra array needed to determine star mask accurately
+    timage = image.copy()
+    timage.array[:, :] = 0
+
+    grng = galsim.BaseDeviate(rng.randint(0, 2**30))
+
+    assert shifts is not None
+    assert mags is not None and len(mags) == len(shifts)
+    assert mask_threshold is not None
+
+    max_bleed_mag = get_max_mag_with_bleed(band=band)
+
+    wcs = image.wcs
+    for obj, shift, mag in zip(objlist, shifts, mags):
+
+        obj = _set_star_gsparams(obj, mag, noise)
+
+        # profiles can have detectably sharp edges if the
+        # profile is very high s/n and we have not set the
+        # thresholds right in the gs params.
+        #
+        # photon shooting reduces these sharp edges, reducing
+        # sensitivity to such an error
+
+        world_pos = coadd_bbox_cen_gs_skypos.deproject(
+            shift.x * galsim.arcsec,
+            shift.y * galsim.arcsec,
+        )
+
+        image_pos = wcs.toImage(world_pos)
+        local_wcs = wcs.local(image_pos=image_pos)
+
+        convolved_object = get_convolved_object(obj, psf, image_pos)
+
+        maxn = 10_000_000
+        n_photons = int(obj.flux) if obj.flux < maxn else maxn
```

Comment:
Thanks for the tip, I'll implement this

Reply:
It looks like you have to give `0` rather than `None`
```python
        # Make sure the type of n_photons is correct and has a valid value:
>       if n_photons < 0.:
E       TypeError: '<' not supported between instances of 'NoneType' and 'float'
``

Reply:
Sorry, yes.  The default is 0, not None.  My mistake.